test/Kestrel.FunctionalTests/AddressRegistrationTests.cs

```diff
@@ -184,10 +195,119 @@ public AddressRegistrationTests(ITestOutputHelper output)
+                })
+                .ConfigureLogging(_configureLoggingDelegate)
+                .Configure(ConfigureEchoAddress);
+
+            using (var host = hostBuilder.Build())
+            {
+                host.Start();
+
+                foreach (var testUrl in testUrls.Select(testUrl => $"{testUrl}:{port}"))
+                {
+                    var response = await HttpClientSlim.GetStringAsync(testUrl, validateCertificate: false);
+
+                    // Compare the response with Uri.ToString(), rather than testUrl directly.
+                    // Required to handle IPv6 addresses with zone index, like "fe80::3%1"
+                    Assert.Equal(new Uri(testUrl).ToString(), response);
+                }
+            }
+        }
+
+        [Fact]
+        public void ThrowsWhenListenLocalhostWithDynamicPort()
+        {
+            var hostBuilder = TransportSelector.GetWebHostBuilder()
+                .UseKestrel(options =>
+                {
+                    options.ListenLocalhost(0);
+                })
+                .Configure(ConfigureEchoAddress);
+
+            using (var host = hostBuilder.Build())
+            {
+                Assert.Throws<InvalidOperationException>(() => host.Start());
+            }
+        }
+
+        [ConditionalFact]
+        [NetworkIsReachable]
+        public async Task ListenPrefix_HostName_Success()
+        {
+            var hostName = Dns.GetHostName();
+            await ListenPrefix_StaticPort_Success($"http://{hostName}:0", $"http://{hostName}");
+        }
+
+        [Theory]
+        [MemberData(nameof(AddressRegistrationDataIPv4))]
+        public async Task ListenPrefix_IPv4_Success(string addressInput, string testUrl)
+        {
+            await ListenPrefix_StaticPort_Success(addressInput, testUrl);
+        }
+
+        [ConditionalTheory]
+        [MemberData(nameof(AddressRegistrationDataIPv6))]
+        [IPv6SupportedCondition]
+        public async Task ListenPrefix_IPv6_Success(string addressInput, string[] testUrls)
+        {
+            await ListenPrefix_StaticPort_Success(addressInput, testUrls);
+        }
+
+        private Task ListenPrefix_StaticPort_Success(string prefix, string testUrl) => ListenPrefix_StaticPort_Success(prefix, new[] { testUrl });
+        
+        private async Task ListenPrefix_Success(string prefix, string[] testUrls, int port = 0)
+        {
+            var hostBuilder = TransportSelector.GetWebHostBuilder()
+                .UseKestrel(options =>
+                {
+                    options.ListenPrefix(prefix + ":" + port);
+                })
+                .ConfigureLogging(_configureLoggingDelegate)
+                .Configure(ConfigureEchoAddress);
+
+            using (var host = hostBuilder.Build())
+            {
+                host.Start();
+
+                foreach (var testUrl in testUrls.Select(testUrl => $"{testUrl}:{port == 0 ? host.GetPort() : port}"))
```

Comment:
You have to wrap the ternary expression in parens. Otherwise the compiler thinks the `:` is a format specifier delimiter.